epics/pv.py

```diff
@@ -356,10 +356,22 @@ def _set_charval(self, val, call_ca=True):
         # char waveform as string
         if ntype == dbr.CHAR and self.count < ca.AUTOMONITOR_MAXLENGTH:
             if isinstance(val, ca.numpy.ndarray):
+                # a numpy array
                 val = val.tolist()
-            elif self.count==1: # handles single character in waveform
-                val = [val]
-            val = list(val)
+
+                if not isinstance(val, list):
+                    # a scalar value from numpy, tolist() turns it into a
+                    # native python integer
+                    val = [val.tolist()]
+            else:
+                try:
+                    # otherwise, try forcing it into a list. this will fail for
+                    # scalar types
+                    val = list(val)
```

Comment:
`list()` here should work in the case of a ctypes array:
```python
In [2]: byte_array = (ctypes.c_ubyte * 5)()

In [3]: list(byte_array)
Out[3]: [0, 0, 0, 0, 0]
```

but fail in the case of a scalar value:
```python
In [6]: list(ctypes.c_ubyte())   # raises TypeError
```